applications/GeoMechanicsApplication/custom_elements/compressibility_calculator.cpp

```diff
@@ -0,0 +1,87 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Richard Faasse
+//
+#include "compressibility_calculator.h"
+#include "custom_utilities/transport_equation_utilities.hpp"
+#include "geo_mechanics_application_variables.h"
+
+namespace Kratos
+{
+
+CompressibilityCalculator::CompressibilityCalculator(InputProvider rInputProvider)
+    : mInputProvider(std::move(rInputProvider))
+{
+}
+
+Matrix CompressibilityCalculator::LHSContribution()
+{
+    return LHSContribution(CalculateCompressibilityMatrix());
+}
+
+Vector CompressibilityCalculator::RHSContribution()
+{
+    return RHSContribution(CalculateCompressibilityMatrix());
+}
+
+Vector CompressibilityCalculator::RHSContribution(const Matrix& rCompressibilityMatrix) const
+{
+    return -prod(rCompressibilityMatrix, mInputProvider.GetNodalValues(DT_WATER_PRESSURE));
+}
+
+Matrix CompressibilityCalculator::LHSContribution(const Matrix& rCompressibilityMatrix) const
+{
+    return rCompressibilityMatrix * mInputProvider.GetDtPressureCoefficient();
+}
+
+std::pair<Matrix, Vector> CompressibilityCalculator::CalculateLeftAndRightHandSide()
+{
+    const auto compressibility_matrix = CalculateCompressibilityMatrix();
+    return {(LHSContribution(compressibility_matrix)), (RHSContribution(compressibility_matrix))};
+}
+
+Matrix CompressibilityCalculator::CalculateCompressibilityMatrix() const
+{
+    const auto& r_N_container            = mInputProvider.GetNContainer();
+    const auto& integration_coefficients = mInputProvider.GetIntegrationCoefficients();
+    auto        result = Matrix{ZeroMatrix{r_N_container.size2(), r_N_container.size2()}};
+    for (unsigned int integration_point_index = 0;
```

Comment:
Can we get rid of the raw loop?

Reply:
Hmm... that won't be easy. The problem is that you're iterating three "ranges": the rows of `r_N_container`, the retention laws, and `integration_coefficients`. The STL algorithms that are available in C++ 17 are intended for either single ranges or two ranges. The only option that I see is to use Boost Range, which offers a way to iterate more than two ranges simultaneously (which requires Boost 1.56+). However, it won't be easy to use it in this case, let alone that it will result in an easy-to-read piece of code.

Reply:
Okay, than we'll leave it as is